src/binutils/doc/size.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man 5.01 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "SIZE 1"
.TH SIZE 1 2026-06-03 binutils-2.16.91 "GNU Development Tools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
size \- list section sizes and total size.
.SH SYNOPSIS
.IX Header "SYNOPSIS"
size [\fB\-A\fR|\fB\-B\fR|\fB\-\-format=\fR\fIcompatibility\fR]
     [\fB\-\-help\fR]
     [\fB\-d\fR|\fB\-o\fR|\fB\-x\fR|\fB\-\-radix=\fR\fInumber\fR]
     [\fB\-t\fR|\fB\-\-totals\fR]
     [\fB\-\-target=\fR\fIbfdname\fR] [\fB\-V\fR|\fB\-\-version\fR]  
     [\fIobjfile\fR...]
.SH DESCRIPTION
.IX Header "DESCRIPTION"
The GNU \fBsize\fR utility lists the section sizes\-\-\-and the total
size\-\-\-for each of the object or archive files \fIobjfile\fR in its
argument list.  By default, one line of output is generated for each
object file or each module in an archive.
.PP
\&\fIobjfile\fR... are the object files to be examined.
If none are specified, the file \f(CW\*(C`a.out\*(C'\fR will be used.
.SH OPTIONS
.IX Header "OPTIONS"
The command line options have the following meanings:
.IP \fB\-A\fR 4
.IX Item "-A"
.PD 0
.IP \fB\-B\fR 4
.IX Item "-B"
.IP \fB\-\-format=\fR\fIcompatibility\fR 4
.IX Item "--format=compatibility"
.PD
Using one of these options, you can choose whether the output from GNU
\&\fBsize\fR resembles output from System V \fBsize\fR (using \fB\-A\fR,
or \fB\-\-format=sysv\fR), or Berkeley \fBsize\fR (using \fB\-B\fR, or
\&\fB\-\-format=berkeley\fR).  The default is the one-line format similar to
Berkeley's.
.Sp
Here is an example of the Berkeley (default) format of output from
\&\fBsize\fR:
.Sp
.Vb 4
\&        $ size \-\-format=Berkeley ranlib size
\&        text    data    bss     dec     hex     filename
\&        294880  81920   11592   388392  5ed28   ranlib
\&        294880  81920   11888   388688  5ee50   size
.Ve
.Sp
This is the same data, but displayed closer to System V conventions:
.Sp
.Vb 7
\&        $ size \-\-format=SysV ranlib size
\&        ranlib  :
\&        section         size         addr
\&        .text         294880         8192       
\&        .data          81920       303104       
\&        .bss           11592       385024       
\&        Total         388392    
\&        
\&        
\&        size  :
\&        section         size         addr
\&        .text         294880         8192       
\&        .data          81920       303104       
\&        .bss           11888       385024       
\&        Total         388688
.Ve
.IP \fB\-\-help\fR 4
.IX Item "--help"
Show a summary of acceptable arguments and options.
.IP \fB\-d\fR 4
.IX Item "-d"
.PD 0
.IP \fB\-o\fR 4
.IX Item "-o"
.IP \fB\-x\fR 4
.IX Item "-x"
.IP \fB\-\-radix=\fR\fInumber\fR 4
.IX Item "--radix=number"
.PD
Using one of these options, you can control whether the size of each
section is given in decimal (\fB\-d\fR, or \fB\-\-radix=10\fR); octal
(\fB\-o\fR, or \fB\-\-radix=8\fR); or hexadecimal (\fB\-x\fR, or
\&\fB\-\-radix=16\fR).  In \fB\-\-radix=\fR\fInumber\fR, only the three
values (8, 10, 16) are supported.  The total size is always given in two
radices; decimal and hexadecimal for \fB\-d\fR or \fB\-x\fR output, or
octal and hexadecimal if you're using \fB\-o\fR.
.IP \fB\-t\fR 4
.IX Item "-t"
.PD 0
.IP \fB\-\-totals\fR 4
.IX Item "--totals"
.PD
Show totals of all objects listed (Berkeley format listing mode only).
.IP \fB\-\-target=\fR\fIbfdname\fR 4
.IX Item "--target=bfdname"
Specify that the object-code format for \fIobjfile\fR is
\&\fIbfdname\fR.  This option may not be necessary; \fBsize\fR can
automatically recognize many formats.
.IP \fB\-V\fR 4
.IX Item "-V"
.PD 0
.IP \fB\-\-version\fR 4
.IX Item "--version"
.PD
Display the version number of \fBsize\fR.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBar\fR\|(1), \fBobjdump\fR\|(1), \fBreadelf\fR\|(1), and the Info entries for \fIbinutils\fR.
.SH COPYRIGHT
.IX Header "COPYRIGHT"
Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999,
2000, 2001, 2002, 2003, 2004, 2005 Free Software Foundation, Inc.
.PP
Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU Free Documentation License, Version 1.1
or any later version published by the Free Software Foundation;
with no Invariant Sections, with no Front-Cover Texts, and with no
Back-Cover Texts.  A copy of the license is included in the
section entitled ``GNU Free Documentation License''.


ChangeLog
2026-05-24  Eric Gallager  <1388660+cooljeanius@users.noreply.github.com>

	Merge pull request #202 from zaidanfaiz/fix-cgen-guile3-conditional-defines

	Fix CGEN conditional definitions for Guile 3

2026-05-24  Eric Gallager  <egall@gwmail.gwu.edu>

	Testing cgen subdir Makefile targets

	The following Makefile targets work in the cgen subdir with --target=arm-apple-darwin:
	- html-local
	- sim-arch
	The following ones fail:
	- opcodes
	- sim-cpu
	- gas-test
	- sim-test

	Change-Id: I32144ef234c439c59e97d6b6bbfcf9da7bcf2764

2026-05-24  Eric Gallager  <1388660+cooljeanius@users.noreply.github.com>

	Merge pull request #203 from cooljeanius/dependabot/npm_and_yarn/src/texinfo/js/npm-07f3ad912e

	Bump the npm group across 1 directory with 2 updates

2026-05-24  dependabot[bot]  <49699333+dependabot[bot]@users.noreply.github.com>

	Bump the npm group across 1 directory with 2 updates

	Bumps the npm group with 2 updates in the /src/texinfo/js directory: [dompurify](https://github.com/cure53/DOMPurify) and [eslint](https://github.com/eslint/eslint).


	Updates `dompurify` from 3.4.2 to 3.4.3
	- [Release notes](https://github.com/cure53/DOMPurify/releases)
	- [Commits](https://github.com/cure53/DOMPurify/compare/3.4.2...3.4.3)

	Updates `eslint` from 10.3.0 to 10.4.0
	- [Release notes](https://github.com/eslint/eslint/releases)
	- [Commits](https://github.com/eslint/eslint/compare/v10.3.0...v10.4.0)

	---
	updated-dependencies:
	- dependency-name: dompurify
	  dependency-version: 3.4.3
	  dependency-type: direct:production
	  update-type: version-update:semver-patch
	  dependency-group: npm
	- dependency-name: eslint
	  dependency-version: 10.4.0
	  dependency-type: direct:development
	  update-type: version-update:semver-minor
	  dependency-group: npm
	...

2026-05-24  Eric Gallager  <1388660+cooljeanius@users.noreply.github.com>

	Merge pull request #201 from cooljeanius/dependabot/github_actions/actions-0f1baada16

	Bump the actions group across 1 directory with 4 updates

2026-05-23  dependabot[bot]  <49699333+dependabot[bot]@users.noreply.github.com>

	Bump the actions group across 1 directory with 4 updates

	Bumps the actions group with 4 updates in the / directory: [actions/checkout](https://github.com/actions/checkout), [mozilla-actions/sccache-action](https://github.com/mozilla-actions/sccache-action), [actions/upload-artifact](https://github.com/actions/upload-artifact) and [github/codeql-action](https://github.com/github/codeql-action).


	Updates `actions/checkout` from 4 to 6
	- [Release notes](https://github.com/actions/checkout/releases)
	- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
	- [Commits](https://github.com/actions/checkout/compare/v4...v6)

	Updates `mozilla-actions/sccache-action` from 0.0.9 to 0.0.10
	- [Release notes](https://github.com/mozilla-actions/sccache-action/releases)
	- [Commits](https://github.com/mozilla-actions/sccache-action/compare/v0.0.9...v0.0.10)

	Updates `actions/upload-artifact` from 4 to 7
	- [Release notes](https://github.com/actions/upload-artifact/releases)
	- [Commits](https://github.com/actions/upload-artifact/compare/v4...v7)

	Updates `github/codeql-action` from 4.35.3 to 4.35.5
	- [Release notes](https://github.com/github/codeql-action/releases)
	- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
	- [Commits](https://github.com/github/codeql-action/compare/v4.35.3...v4.35.5)

	---
	updated-dependencies:
	- dependency-name: actions/checkout
	  dependency-version: '6'
	  dependency-type: direct:production
	  update-type: version-update:semver-major
	  dependency-group: actions
	- dependency-name: actions/upload-artifact
	  dependency-version: '7'
	  dependency-type: direct:production
	  update-type: version-update:semver-major
	  dependency-group: actions
	- dependency-name: github/codeql-action
	  dependency-version: 4.35.4
	  dependency-type: direct:production
	  update-type: version-update:semver-patch
	  dependency-group: actions
	- dependency-name: mozilla-actions/sccache-action
	  dependency-version: 0.0.10
	  dependency-type: direct:production
	  update-type: version-update:semver-patch
	  dependency-group: actions
	...

2026-05-16  Codex  <codex@openai.com>

	Fix CGEN conditional definitions for Guile 3

2026-05-16  Eric Gallager  <1388660+cooljeanius@users.noreply.github.com>

	Merge pull request #200 from cooljeanius/dependabot/gradle/gradle-2978c64675

	Bump com.google.guava:guava from 33.5.0-jre to 33.6.0-jre in the gradle group

2026-05-16  dependabot[bot]  <49699333+dependabot[bot]@users.noreply.github.com>

	Bump com.google.guava:guava in the gradle group

	Bumps the gradle group with 1 update: [com.google.guava:guava](https://github.com/google/guava).


	Updates `com.google.guava:guava` from 33.5.0-jre to 33.6.0-jre
	- [Release notes](https://github.com/google/guava/releases)
	- [Commits](https://github.com/google/guava/commits)

	---
	updated-dependencies:
	- dependency-name: com.google.guava:guava
	  dependency-version: 33.6.0-jre
	  dependency-type: direct:production
	  update-type: version-update:semver-minor
	  dependency-group: gradle
	...

2026-05-12  Eric Gallager  <egall@gwmail.gwu.edu>

	shuffle workflows around a bit more

	Change-Id: I8a9136fcca0d7f0ed02238a70c0ef212155bb25b

2026-05-12  Eric Gallager  <1388660+cooljeanius@users.noreply.github.com>

	Merge pull request #199 from cooljeanius/macos_arm

	Add aarch64-apple-darwin support

2026-05-12  Eric Gallager  <egall@gwmail.gwu.edu>

	noticed things during review

	Change-Id: I3c14c7cdbb1963b16a9e8490a10b896e0a342e62

2026-05-12  Eric Gallager  <egall@gwmail.gwu.edu>

	Merge branch 'master' into macos_arm

	Change-Id: Ic0f04a827eb5c57bbd7de9e39342f69dc5895572

2026-05-12  Eric Gallager  <egall@gwmail.gwu.edu>

	a few minor workflow updates

	Change-Id: Ie69cce10eb05a518ad49389ab260d80080387185

2026-05-12  Eric Gallager  <egall@gwmail.gwu.edu>

	more frustrations, mig and (separately) gettext


src/binutils/size.c
/* size.c -- report size of various sections of an executable file.
   Copyright 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000,
   2001, 2002, 2003, 2004, 2005 Free Software Foundation, Inc.

   This file is part of GNU Binutils.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St., 5th Floor, Boston, MA 02110-1301, USA */

/* Extensions/incompatibilities:
   o - BSD output has filenames at the end.
   o - BSD output can appear in different radicies.
   o - SysV output has less redundant whitespace.  Filename comes at end.
   o - SysV output doesn't show VMA which is always the same as the PMA.
   o - We also handle core files.
   o - We also handle archives.
   If you write shell scripts which manipulate this info then you may be
   out of luck; there is no --compatibility or --pedantic option.  */

#include "bfd.h"
#include "bucomm.h"
#include "libiberty.h"
#include "getopt.h"

#include "sysdep.h"

#ifdef HAVE_STDINT_H
# include <stdint.h>
#endif /* HAVE_STDINT_H */

#ifndef BSD_DEFAULT
# define BSD_DEFAULT 1
#endif /* !BSD_DEFAULT */

/* Program options: */
enum program_options {
    decimal, octal, hex
} radix = decimal;

/* 0 means use AT&T-style output: */
static int berkeley_format = BSD_DEFAULT;

extern int show_version;
int show_version = 0;
extern int show_help;
int show_help = 0;
extern int show_totals;
int show_totals = 0;

static bfd_size_type total_bsssize;
static bfd_size_type total_datasize;
static bfd_size_type total_textsize;

/* Program exit status: */
extern int return_code;
int return_code = 0;

static char *target = NULL;

/* Static declarations: */
static ATTRIBUTE_NORETURN void usage(FILE *, int);
static void display_file(const char *);
static void display_bfd(bfd *);
static void display_archive(bfd *);
static size_t size_number(bfd_size_type);
static void rprint_number(int8_t, bfd_size_type);
static void print_berkeley_format(bfd *);
static void sysv_internal_sizer(bfd *, asection *, void *);
static void sysv_internal_printer(bfd *, asection *, void *);
static void print_sysv_format(bfd *);
static void print_sizes(bfd *file);
static void berkeley_sum(bfd *, sec_ptr, void *);

static ATTRIBUTE_NORETURN void
usage(FILE *stream, int status)
{
  fprintf(stream, _("Usage: %s [option(s)] [file(s)]\n"), program_name);
  fprintf(stream, _(" Displays the sizes of sections inside binary files\n"));
  fprintf(stream, _(" If no input file(s) are specified, a.out is assumed\n"));
  fprintf(stream, _(" The options are:\n\
  -A|-B     --format={sysv|berkeley}  Select output style (default is %s)\n\
  -o|-d|-x  --radix={8|10|16}         Display numbers in octal, decimal or hex\n\
  -t        --totals                  Display the total sizes (Berkeley only)\n\
            --target=<bfdname>        Set the binary file format\n\
  -h        --help                    Display this information\n\
  -v        --version                 Display the program's version\n\
\n"),
#if BSD_DEFAULT
  "berkeley"
#else
  "sysv"
#endif /* BSD_DEFAULT */
);
  list_supported_targets(program_name, stream);
  if (status == 0)
    fprintf(stream, _("Report bugs to %s\n"), REPORT_BUGS_TO);
  xexit(status);
}

static struct option long_options[] =
{
  { "format", required_argument, 0, 200 },
  { "radix", required_argument, 0, 201 },
  { "target", required_argument, 0, 202 },
  { "totals", no_argument, &show_totals, 1 },
  { "version", no_argument, &show_version, 1 },
  { "help", no_argument, &show_help, 1 },
  { 0, no_argument, 0, 0 }
};

int main(int, char **);

int
main(int argc, char **argv)
{
  long temp;
  int c;

#if defined(HAVE_SETLOCALE) && defined(HAVE_LC_MESSAGES)
  setlocale(LC_MESSAGES, "");
#endif /* HAVE_SETLOCALE && HAVE_LC_MESSAGES */
#if defined(HAVE_SETLOCALE)
  setlocale(LC_CTYPE, "");
#endif /* HAVE_SETLOCALE */
  bindtextdomain(PACKAGE, LOCALEDIR);
  textdomain(PACKAGE);

  program_name = *argv;
  xmalloc_set_program_name(program_name);

  bfd_init();
  set_default_bfd_target();

  while ((c = getopt_long(argc, argv, "ABHhVvdfotx", long_options,
                          (int *)0)) != EOF)
    switch (c)
      {
      case 200:		/* --format */
	switch (*optarg)
	  {
	  case 'B':
	  case 'b':
	    berkeley_format = 1;
	    break;
	  case 'S':
	  case 's':
	    berkeley_format = 0;
	    break;
	  default:
	    non_fatal(_("invalid argument to --format: %s"), optarg);
	    usage(stderr, 1);
	  }
	break;

      case 202:		/* --target */
	target = optarg;
	break;

      case 201:		/* --radix */
#ifdef ANSI_LIBRARIES
	temp = strtol(optarg, NULL, 10);
#else
	temp = atol(optarg);
#endif /* ANSI_LIBRARIES */
	switch (temp)
	  {
	  case 10:
	    radix = decimal;
	    break;
	  case 8:
	    radix = octal;
	    break;
	  case 16:
	    radix = hex;
	    break;
	  default:
	    non_fatal(_("Invalid radix: %s\n"), optarg);
	    usage(stderr, 1);
	  }
	break;

      case 'A':
	berkeley_format = 0;
	break;
      case 'B':
	berkeley_format = 1;
	break;
      case 'v':
      case 'V':
	show_version = 1;
	break;
      case 'd':
	radix = decimal;
	break;
      case 'x':
	radix = hex;
	break;
      case 'o':
	radix = octal;
	break;
      case 't':
	show_totals = 1;
	break;
      case 'f': /* FIXME : For sysv68, `-f' means `full format', i.e.
		   `[fname:] M(.text) + N(.data) + O(.bss) + P(.comment) = Q'
		   where `fname: ' appears only if there are >= 2 input files,
		   and M, N, O, P, Q are expressed in decimal by default,
		   hexa or octal if requested by `-x' or `-o'.
		   Just to make things interesting, Solaris also accepts -f,
		   which prints out the size of each allocatable section, the
		   name of the section, and the total of the section sizes.  */
		/* For the moment, accept `-f' silently, and ignore it.  */
	break;
      case 0:
	break;
      case 'h':
      case 'H':
      case '?':
	usage(stderr, 1);
      default:
        break;
      }

  if (show_version) {
    print_version("size");
  }
  if (show_help) {
    usage(stdout, 0);
  }

  if (optind == argc) {
    display_file("a.out");
  } else {
    for (; optind < argc;) {
      display_file(argv[optind++]);
    }
  }

  if (show_totals && berkeley_format)
    {
      bfd_size_type total;
      total = (total_textsize + total_datasize + total_bsssize);

      rprint_number(7, total_textsize);
      putchar('\t');
      rprint_number(7, total_datasize);
      putchar('\t');
      rprint_number(7, total_bsssize);
      printf(((radix == octal) ? "\t%7lo\t%7lx\t" : "\t%7lu\t%7lx\t"),
             (unsigned long)total, (unsigned long)total);
      fputs("(TOTALS)\n", stdout);
    }

  return return_code;
}

/* Display stats on file or archive member ABFD: */
static void
display_bfd(bfd *abfd)
{
  char **matching;

  if (bfd_check_format(abfd, bfd_archive)) {
    /* An archive within an archive: */
    return;
  }

  if (bfd_check_format_matches(abfd, bfd_object, &matching))
    {
      print_sizes(abfd);
      printf("\n");
      return;
    }

  if (bfd_get_error() == bfd_error_file_ambiguously_recognized)
    {
      bfd_nonfatal(bfd_get_filename(abfd));
      list_matching_formats(matching);
      free(matching);
      return_code = 3;
      return;
    }

  if (bfd_check_format_matches(abfd, bfd_core, &matching))
    {
      const char *core_cmd;

      print_sizes(abfd);
      fputs(" (core file", stdout);

      core_cmd = bfd_core_file_failing_command(abfd);
      if (core_cmd) {
	printf(" invoked as %s", core_cmd);
      }

      puts(")\n");
      return;
    }

  bfd_nonfatal(bfd_get_filename(abfd));

  if (bfd_get_error() == bfd_error_file_ambiguously_recognized)
    {
      list_matching_formats(matching);
      free(matching);
    }

  return_code = 3;
}

static void
display_archive(bfd *file)
{
  bfd *arfile = (bfd *)NULL;
  bfd *last_arfile = (bfd *)NULL;

  for (;;)
    {
      bfd_set_error(bfd_error_no_error);

      arfile = bfd_openr_next_archived_file(file, arfile);
      if (arfile == NULL)
	{
	  if (bfd_get_error() != bfd_error_no_more_archived_files)
	    {
	      bfd_nonfatal(bfd_get_filename(file));
	      return_code = 2;
	    }
	  break;
	}

      display_bfd(arfile);

      if (last_arfile != NULL) {
	bfd_close(last_arfile);
      }
      last_arfile = arfile;
    }

  if (last_arfile != NULL) {
    bfd_close(last_arfile);
  }
}

/* */
static void
display_file(const char *filename)
{
  bfd *file;

  if (get_file_size(filename) < 1) {
    return;
  }

  file = bfd_openr(filename, target);
  if (file == NULL)
    {
      bfd_nonfatal(filename);
      return_code = 1;
      return;
    }

  if (bfd_check_format(file, bfd_archive)) {
    display_archive(file);
  } else {
    display_bfd(file);
  }

  if (!bfd_close(file))
    {
      bfd_nonfatal(filename);
      return_code = 1;
      return;
    }
}

#if defined(__GNUC__)
# ifdef sprintf
#  undef sprintf
# endif /* sprintf */
# pragma GCC poison sprintf
#endif /* __GNUC__ */

/* This is what lexical functions are for: */
static size_t
size_number(bfd_size_type num)
{
  char buffer[40];

  snprintf(buffer, sizeof(buffer),
	   ((radix == decimal) ? "%lu" :
	    ((radix == octal) ? "0%lo" : "0x%lx")),
	   (unsigned long)num);

  return strlen(buffer);
}

static void
rprint_number(int8_t width, bfd_size_type num)
{
  char buffer[40];

  snprintf(buffer, sizeof(buffer),
	   ((radix == decimal) ? "%lu" :
	    ((radix == octal) ? "0%lo" : "0x%lx")),
	   (unsigned long)num);

  printf("%*s", width, buffer);
}

static bfd_size_type bsssize;
static bfd_size_type datasize;
static bfd_size_type textsize;

static void
berkeley_sum(bfd *abfd ATTRIBUTE_UNUSED, sec_ptr sec,
             void *ignore ATTRIBUTE_UNUSED)
{
  flagword flags;
  bfd_size_type size;

  flags = bfd_get_section_flags(abfd, sec);
  if ((flags & SEC_ALLOC) == 0) {
    return;
  }

  size = bfd_get_section_size(sec);
  if (((flags & SEC_CODE) != 0) || ((flags & SEC_READONLY) != 0)) {
    textsize += size;
  } else if ((flags & SEC_HAS_CONTENTS) != 0) {
    datasize += size;
  } else {
    bsssize += size;
  }
}

static void
print_berkeley_format(bfd *abfd)
{
  static int files_seen = 0;
  bfd_size_type total;

  bsssize = 0;
  datasize = 0;
  textsize = 0;

  bfd_map_over_sections(abfd, berkeley_sum, NULL);

  if (files_seen++ == 0) {
    puts((radix == octal)
         ? "   text\t   data\t    bss\t    oct\t    hex\tfilename"
         : "   text\t   data\t    bss\t    dec\t    hex\tfilename");
  }

  total = textsize + datasize + bsssize;

  if (show_totals)
    {
      total_textsize += textsize;
      total_datasize += datasize;
      total_bsssize  += bsssize;
    }

  rprint_number(7, textsize);
  putchar('\t');
  rprint_number(7, datasize);
  putchar('\t');
  rprint_number(7, bsssize);
  printf(((radix == octal) ? "\t%7lo\t%7lx\t" : "\t%7lu\t%7lx\t"),
         (unsigned long)total, (unsigned long)total);

  fputs(bfd_get_filename(abfd), stdout);

  if (bfd_my_archive(abfd)) {
    printf(" (ex %s)", bfd_get_filename(bfd_my_archive(abfd)));
  }
}

/* I REALLY miss lexical functions! */
/* (why though?) */
extern bfd_size_type svi_total;
bfd_size_type svi_total = 0;
extern bfd_vma svi_maxvma;
bfd_vma svi_maxvma = 0;
extern size_t svi_namelen;
size_t svi_namelen = 0;
extern size_t svi_vmalen;
size_t svi_vmalen = 0;
extern size_t svi_sizelen;
size_t svi_sizelen = 0;

static void
sysv_internal_sizer(bfd *file ATTRIBUTE_UNUSED, sec_ptr sec,
                    void *ignore ATTRIBUTE_UNUSED)
{
  bfd_size_type size = bfd_section_size(file, sec);

  if (! bfd_is_abs_section(sec)
      && ! bfd_is_com_section(sec)
      && ! bfd_is_und_section(sec))
    {
      size_t namelen = strlen(bfd_section_name(file, sec));

      if (namelen > svi_namelen) {
	svi_namelen = namelen;
      }

      svi_total += size;

      if (bfd_section_vma(file, sec) > svi_maxvma) {
	svi_maxvma = bfd_section_vma(file, sec);
      }
    }
}

static void
sysv_internal_printer(bfd *file ATTRIBUTE_UNUSED, sec_ptr sec,
                      void *ignore ATTRIBUTE_UNUSED)
{
  bfd_size_type size = bfd_section_size(file, sec);

  if (! bfd_is_abs_section(sec)
      && ! bfd_is_com_section(sec)
      && ! bfd_is_und_section(sec))
    {
      svi_total += size;

      printf("%-*s   ", (int8_t)svi_namelen, bfd_section_name(file, sec));
      rprint_number((int8_t)svi_sizelen, size);
      printf("   ");
      rprint_number((int8_t)svi_vmalen, bfd_section_vma(file, sec));
      printf("\n");
    }
}

static void
print_sysv_format(bfd *file)
{
  /* Size all of the columns: */
  svi_total = 0;
  svi_maxvma = 0;
  svi_namelen = 0;
  bfd_map_over_sections(file, sysv_internal_sizer, NULL);
  svi_vmalen = size_number((bfd_size_type)svi_maxvma);

  if ((size_t)svi_vmalen < (sizeof("addr") - 1)) {
    svi_vmalen = (sizeof("addr") - 1);
  }

  svi_sizelen = size_number(svi_total);
  if ((size_t)svi_sizelen < (sizeof("size") - 1)) {
    svi_sizelen = (sizeof("size") - 1);
  }

  svi_total = 0;
  printf("%s  ", bfd_get_filename(file));

  if (bfd_my_archive(file)) {
    printf(" (ex %s)", bfd_get_filename(bfd_my_archive(file)));
  }

  printf(":\n%-*s   %*s   %*s\n", (int8_t)svi_namelen, "section",
         (int8_t)svi_sizelen, "size", (int8_t)svi_vmalen, "addr");

  bfd_map_over_sections(file, sysv_internal_printer, NULL);

  printf("%-*s   ", (int8_t)svi_namelen, "Total");
  rprint_number((int8_t)svi_sizelen, svi_total);
  printf("\n\n");
}

static void
print_sizes(bfd *file)
{
  if (berkeley_format) {
    print_berkeley_format(file);
  } else {
    print_sysv_format(file);
  }
}

/* EOF */
